### PR TITLE
feat(backend-mavericks): Implement application user authentication.

### PR DESCRIPTION
**Pull request by bahailu-abera**
Implementing the API endpoints for user account creation, account update, and login.

Implementing password hashing and salting for securely storing user credentials in the database.

Implementing token-based authentication to protect API endpoints and reject unauthorized access attempts.

Implementing user authentication middle-ware to ensure authenticated access to protected endpoints.

Ensuring that all API endpoints return appropriate HTTP status codes and error messages for invalid requests.

Writing unit to thoroughly test all API endpoints and ensure that they function correctly and as expected.

### CODE REVIEW

**Comment by danielteferadeti** [comment]
In the description section the PR make sure to discripe what you have done in this task. Make sure to add what you have covered, any assumtions you have taken and any thing necessary a reviewer need to know to review you code.

File: Backend/.gitignore
Comment (danielteferadeti) [inline Backend/.gitignore:1]: This shouldn't have been changed. You manualy added the removed lines at the end of the file and do on emore commit

File: Backend/starter_project_dotnetforce7/BlogApp2/BlogApp.Persistence/Migrations/20230501123711_initialMigration.Designer.cs
Comment (danielteferadeti) [inline Backend/starter_project_dotnetforce7/BlogApp2/BlogApp.Persistence/Migrations/20230501123711_initialMigration.Designer.cs:1]: This migration files should be ignored....ask me(dani) on how we can do this in person.

File: Backend/starter_project_zsilentcoders/BlogApp3/BlogApp.Persistence/Migrations/20230501123711_initialMigration.Designer.cs
Comment (danielteferadeti) [inline Backend/starter_project_zsilentcoders/BlogApp3/BlogApp.Persistence/Migrations/20230501123711_initialMigration.Designer.cs:1]: This migration files should be ignored....ask me(dani) on how we can do this in person.

File: Backend/starter_project_mavericks/BlogApp/BlogApp.Persistence/Migrations/20230501123711_initialMigration.Designer.cs
Comment (danielteferadeti) [inline Backend/starter_project_mavericks/BlogApp/BlogApp.Persistence/Migrations/20230501123711_initialMigration.Designer.cs:1]: MIgration files should be ignored

File: Backend/starter_project_mavericks/BlogApp/BlogApp.Application/Features/Authentication/DTO/SignInResponse.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace BlogApp.Application.Features.Authentication.DTO
+{
+    public class SignInResponse : SignUpResponse
+    {
+        public string Token { get; set; }
+    }
```
Comment (danielteferadeti) [inline Backend/starter_project_mavericks/BlogApp/BlogApp.Application/Features/Authentication/DTO/SignInResponse.cs:12]: on our signin response we can add other properties of user as well if we don't have endpoint that exposes that: this are Id, Username, Firstname ,Lastname and Email of the user.

File: Backend/starter_project_mavericks/BlogApp/BlogApp.Application/Features/Authentication/DTO/SignUpResponse.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+using System.Threading.Tasks;
+
+namespace BlogApp.Application.Features.Authentication.DTO
+{
+    public class SignUpResponse
+    {
+        public string Id { get; set; }
+        public string FirstName { get; set; }
+        public string LastName { get; set; }
+        public string Email { get; set; }
```
Comment (danielteferadeti) [inline Backend/starter_project_mavericks/BlogApp/BlogApp.Application/Features/Authentication/DTO/SignUpResponse.cs:14]: This looks good

File: Backend/starter_project_mavericks/BlogApp/BlogApp.Identity/Repositories/AuthRepository.cs [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+                        if (result.Succeeded)
+                        {
+                            await _userManager.AddToRoleAsync(user, "User");
+                            var createdUser = await _userManager.FindByNameAsync(user.UserName);
+                            var response = _mapper.Map<SignUpResponse>(createdUser);
+
+                            return response;
+                            await transaction.CommitAsync();
+                        }
+                        throw new BadRequestException("Email already exists");
```
Comment (danielteferadeti) [inline Backend/starter_project_mavericks/BlogApp/BlogApp.Identity/Repositories/AuthRepository.cs:102]: This thrown exception message should be relevant to why _userManager faild to create the user. There is no user by this email couse we have validated that at the start our logic.

File: Backend/starter_project_mavericks/BlogApp/BlogApp.Identity/Configurations/Entities/RoleConfiguration.cs
Comment (danielteferadeti) [inline Backend/starter_project_mavericks/BlogApp/BlogApp.Identity/Configurations/Entities/RoleConfiguration.cs:1]: This section loogs great!

File: Backend/starter_project_mavericks/BlogApp/BlogApp.Identity/Migrations/20230507211749_IdentityMigration.Designer.cs
Comment (danielteferadeti) [inline Backend/starter_project_mavericks/BlogApp/BlogApp.Identity/Migrations/20230507211749_IdentityMigration.Designer.cs:1]: All Migrations files should be ignored.

File: Backend/starter_project_mavericks/BlogApp/tests/BlogApp.UnitTests/User/SignInHandlerTests.cs [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+// using BlogApp.Application.Models.Identity;
+// using BlogApp.Application.Responses;
+// using BlogApp.Domain;
+// using Microsoft.AspNetCore.Identity;
+// using BlogApp.UnitTests.Mocks;
+// using Moq;
+// using System.Collections.Generic;
+// using System.Security.Claims;
+// using System.Threading.Tasks;
+// using Xunit;
```
Comment (danielteferadeti) [inline Backend/starter_project_mavericks/BlogApp/tests/BlogApp.UnitTests/User/SignInHandlerTests.cs:15]: The commented section on this file should be removed.

File: Backend/starter_project_mavericks/BlogApp/tests/BlogApp.UnitTests/User/SignInHandlerTests.cs [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+    {
+        private readonly AuthRepository _authRepository;
+        private readonly Mock<UserManager<User>> _userManagerMock;
+        private readonly Mock<SignInManager<User>> _signInManagerMock;
+
+        public SignInHandlerTests()
+        {
+            var services = new ServiceCollection();
+            var configuration = new ConfigurationBuilder()
+                .AddJsonFile("/home/bahailu/Documents/a2sv/Internship-2023/Backend/starter_project_mavericks/BlogApp/BlogApp.API/appsettings.json", optional: true, reloadOnChange: true)
```
Comment (danielteferadeti) [inline Backend/starter_project_mavericks/BlogApp/tests/BlogApp.UnitTests/User/SignInHandlerTests.cs:132]: THe folder path provided here should relative not absolute

File: Backend/starter_project_mavericks/BlogApp/tests/BlogApp.UnitTests/User/SignUpHandlerTests.cs [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        public AuthServiceTests()
+        {
+            
+            var mapper = new MapperConfiguration(cfg =>
+            {
+                cfg.CreateMap<SignupFormDto, User>();
+                cfg.CreateMap<User, SignUpResponse>();
+            }).CreateMapper();
+            var configuration = new ConfigurationBuilder()
+            .AddJsonFile("/home/bahailu/Documents/a2sv/Internship-2023/Backend/starter_project_mavericks/BlogApp/BlogApp.API/appsettings.json", optional: true, reloadOnChange: true)
```
Comment (danielteferadeti) [inline Backend/starter_project_mavericks/BlogApp/tests/BlogApp.UnitTests/User/SignUpHandlerTests.cs:36]: The same here

File: Backend/.gitignore [hunk 7]
```diff
@@ -609,5 +609,3 @@ msbuild.wrn
 **/net7.0/
 **/bin/
 **/.vs/
-
-**/Migrations/
```
Comment (danielteferadeti) [inline Backend/.gitignore:613]: This shouldn't have been changed. You can manualy add the removed lines at the end of the file and do one more commit

**Review by danielteferadeti — COMMENTED** [review]
Overall it looks great! Nice implementation of concepts we went through!

File: Backend/starter_project_mavericks/BlogApp/tests/BlogApp.UnitTests/User/SignInHandlerTests.cs (hunk 5 above)
Comment (bahailu-abera) [reply Backend/starter_project_mavericks/BlogApp/tests/BlogApp.UnitTests/User/SignInHandlerTests.cs:132]: When I use relative path I am getting the following error 

Error Message:
   System.IO.FileNotFoundException : The configuration file '../../../BlogApp.API/appsettings.json' was not found and is not optional.
  Stack Trace:
     at Microsoft.Extensions.Configuration.FileConfigurationProvider.Load(Boolean reload)
   at Microsoft.Extensions.Configuration.ConfigurationRoot..ctor(IList`1 providers)
   at Microsoft.Extensions.Configuration.ConfigurationBuilder.Build()
   at BlogApp.Identity.UnitTests.AuthServiceTests..ctor() in /home/bahailu/Documents/a2sv/Internship-2023/Backend/starter_project_mavericks/BlogApp/tests/BlogApp.UnitTests/User/SignUpHandlerTests.cs:line 37
   at System.RuntimeType.CreateInstanceDefaultCtor(Boolean publicOnly, Boolean wrapExceptions)

**Pull request closed without merging**